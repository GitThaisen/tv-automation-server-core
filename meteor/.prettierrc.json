{
	"arrowParens": "always",
	"bracketSpacing": true,
	"trailingComma": "none",
	"singleQuote": true,
	"jsxBracketSameLine": true,
	"printWidth": 120,
	"useTabs": true,
	"endOfLine": "lf",
	"semi": false
}
