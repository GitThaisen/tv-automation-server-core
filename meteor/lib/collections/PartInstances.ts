import * as _ from 'underscore'
import { TransformedCollection, MongoSelector, FindOptions } from '../typings/meteor'
import { applyClassToDocument, Time, registerCollection } from '../lib'
import { Meteor } from 'meteor/meteor'
import {
	IBlueprintPartInstance,
	BlueprintRuntimeArguments,
	IBlueprintPartInstanceTimings,
	PartEndState,
} from 'tv-automation-sofie-blueprints-integration'
import { createMongoCollection } from './lib'
import { DBPart, Part } from './Parts'
import { PieceInstance, PieceInstances } from './PieceInstances'
import { Pieces } from './Pieces'

export interface DBPartInstance extends IBlueprintPartInstance {
	rundownId: string

	/** Rank of the take that this PartInstance belongs to */
	takeCount: number

	part: DBPart

	timings: PartInstanceTimings

	/** The time the system played back this part, null if not yet finished playing, in milliseconds.
	 * This is set when Take:ing the next part
	 */
	duration?: number // TODO - this can be replaced with timings?
	/** The end state of the previous part, to allow for bits of this to part to be based on what the previous did/was */
	previousPartEndState?: PartEndState

	// /** if the part is inserted after another (for adlibbing) */
	// afterPartInstance?: string
	/** if the part was dunamically inserted (adlib) */
	dynamicallyInserted?: boolean

	/** Runtime blueprint arguments allows Sofie-side data to be injected into the blueprint for an part */
	runtimeArguments?: BlueprintRuntimeArguments
	/** An part should be marked as `dirty` if the part blueprint has been injected with runtimeArguments */
	dirty?: boolean // TODO - remove?

	/** Once a PartInstance is reset it should no longer be shown to the user. */
	reset?: boolean
}
export interface PartInstanceTimings extends IBlueprintPartInstanceTimings {
	/** The playback offset that was set for the last take */
	playOffset?: Time
}

export class PartInstance implements DBPartInstance {
	/** Whether this PartInstance is a temprorary wrapping of a Part */
	public readonly isTemporary: boolean

	// From IBlueprintPartInstance:
	public part: Part
	public _id: string
	public takeCount: number
	public segmentId: string
	public rundownId: string
	public timings: PartInstanceTimings
	// From DBPart:
	// public startedPlayback?: boolean
	// public stoppedPlayback?: boolean
	public duration?: number
	public previousPartEndState?: PartEndState
	// public afterPartInstance?: string
	public dynamicallyInserted?: boolean
	public runtimeArguments?: BlueprintRuntimeArguments
	public dirty?: boolean // TODO - remove?
	public reset?: boolean

	constructor (document: DBPartInstance, isTemporary?: boolean) {
		_.each(_.keys(document), (key) => {
			this[key] = document[key]
		})
		this.isTemporary = isTemporary === true
		this.part = new Part(document.part)
	}
	getPieceInstances (selector?: MongoSelector<PieceInstance>, options?: FindOptions) {
		if (this.isTemporary) {
			throw new Error('Not implemented') // TODO?
			// const pieces = Pieces.find(
			// 	{
			// 		...selector,
			// 		rundownId: this.rundownId,
			// 		partId: this._id
			// 	},
			// 	{
			// 		sort: { _rank: 1 },
			// 		...options
			// 	}
			// ).fetch()
		} else {
			return PieceInstances.find(
				{
					...selector,
					rundownId: this.rundownId,
					partId: this._id
				},
				{
					sort: { _rank: 1 },
					...options
				}
			).fetch()
		}
	}
	getAllPieceInstances () {
		return this.getPieceInstances()
	}

	// getTimings () {
	// 	// return a chronological list of timing events
	// 	let events: Array<{time: Time, type: string, elapsed: Time}> = []
	// 	_.each(['take', 'takeDone', 'startedPlayback', 'takeOut', 'stoppedPlayback', 'next'], (key) => {
	// 		if (this.timings) {
	// 			_.each(this.timings[key], (t: Time) => {
	// 				events.push({
	// 					time: t,
	// 					type: key,
	// 					elapsed: 0
	// 				})
	// 			})
	// 		}
	// 	})
	// 	let prevEv: any = null
	// 	return _.map(
	// 		_.sortBy(events, e => e.time),
	// 		(ev) => {
	// 			if (prevEv) {
	// 				ev.elapsed = ev.time - prevEv.time
	// 			}
	// 			prevEv = ev
	// 			return ev
	// 		}
	// 	)

	// }
	// getLastTake () {
	// 	if (!this.timings) return undefined

	// 	if (!this.timings.take || this.timings.take.length === 0) return undefined

	// 	return this.timings.take[this.timings.take.length - 1]
	// }
	// getLastStartedPlayback () {
	// 	if (!this.timings) return undefined

	// 	if (!this.timings.startedPlayback || this.timings.startedPlayback.length === 0) return undefined

	// 	return this.timings.startedPlayback[this.timings.startedPlayback.length - 1]
	// }
	// getLastPlayOffset () {
	// 	if (!this.timings) return undefined

	// 	if (!this.timings.playOffset || this.timings.playOffset.length === 0) return undefined

	// 	return this.timings.playOffset[this.timings.playOffset.length - 1]
	// }
}

export function WrapPartToTemporaryInstance (part: DBPart): PartInstance {
	return new PartInstance({
		_id: `${part._id}_tmp_instance`,
		rundownId: part.rundownId,
		segmentId: part.segmentId,
		takeCount: -1, // TODO - is this any good?
		part: new Part(part),
		timings: {}
	}, true)
}

export function FindPartInstanceOrWrapToTemporary (partInstances: PartInstance[], part: DBPart): PartInstance {
	return partInstances.find(instance => instance.part._id === part._id) || WrapPartToTemporaryInstance(part)
}

export const PartInstances: TransformedCollection<PartInstance, DBPartInstance> = createMongoCollection<PartInstance>('partInstances', { transform: (doc) => applyClassToDocument(PartInstance, doc) })
registerCollection('PartInstances', PartInstances)
Meteor.startup(() => {
	if (Meteor.isServer) {
		PartInstances._ensureIndex({
			rundownId: 1,
			segmentId: 1,
			_rank: 1
		})
		PartInstances._ensureIndex({
			rundownId: 1,
			_rank: 1
		})
		PartInstances._ensureIndex({
			rundownId: 1,
			partId: 1,
			_rank: 1
		})
	}
})
