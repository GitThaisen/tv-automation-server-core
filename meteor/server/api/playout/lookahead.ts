import { Meteor } from 'meteor/meteor'
import * as _ from 'underscore'
import { LookaheadMode, Timeline as TimelineTypes, OnGenerateTimelineObj } from 'tv-automation-sofie-blueprints-integration'
import { Studio, MappingExt } from '../../../lib/collections/Studios'
import { TimelineObjGeneric, TimelineObjRundown, fixTimelineId, TimelineObjType } from '../../../lib/collections/Timeline'
import { Part, PartId } from '../../../lib/collections/Parts'
import { Piece, Pieces } from '../../../lib/collections/Pieces'
import { sortPiecesByStart } from './pieces'
import { literal, clone, unprotectString, protectString, makePromise, waitForPromiseAll } from '../../../lib/lib'
import { RundownPlaylistPlayoutData, RundownPlaylist } from '../../../lib/collections/RundownPlaylists'
import { PieceInstance, wrapPieceToInstance, PieceInstancePiece, rewrapPieceToInstance } from '../../../lib/collections/PieceInstances'
import { selectNextPart } from './lib'
import { PartInstanceId, PartInstance } from '../../../lib/collections/PartInstances'
import { Rundown, RundownId } from '../../../lib/collections/Rundowns';

const LOOKAHEAD_OBJ_PRIORITY = 0.1

export function getLookeaheadObjects (playoutData: RundownPlaylistPlayoutData, studio: Studio): Array<TimelineObjGeneric> {
	const timelineObjs: Array<TimelineObjGeneric> = []
	const mutateAndPushObject = (rawObj: TimelineObjRundown, i: string, enable: TimelineObjRundown['enable'], mapping: MappingExt, priority: number) => {
		const obj: TimelineObjGeneric = clone(rawObj)

		obj.id = `lookahead_${i}_${obj.id}`
		obj.priority = priority
		obj.enable = enable
		obj.isLookahead = true
		delete obj.keyframes
		delete obj.inGroup // force it to be cleared

		if (mapping.lookahead === LookaheadMode.PRELOAD) {
			obj.lookaheadForLayer = obj.layer
			obj.layer += '_lookahead'
		}

		timelineObjs.push(obj)
	}

	const calculateStartAfterPreviousObj = (prevObj: TimelineObjRundown): TimelineTypes.TimelineEnable => {
		const prevHasDelayFlag = (prevObj.classes || []).indexOf('_lookahead_start_delay') !== -1

		// Start with previous piece
		const startOffset = prevHasDelayFlag ? 2000 : 0
		return {
			start: `#${prevObj.id}.start + ${startOffset}`
		}
	}

	// Get all the parts and figure out which ones to consider for future lookahead
	const orderedParts = playoutData.rundownPlaylist.getAllOrderedParts()
	const nextNextPart = selectNextPart(playoutData.nextPartInstance || playoutData.currentPartInstance || null, orderedParts)
	let futureParts: Part[] = []
	if (nextNextPart) {
		futureParts = orderedParts.slice(nextNextPart.index)
	}

	const ps = _.map(studio.mappings || {}, (mapping: MappingExt, layerId: string) => {
		// This runs some db queries, so lets try and run in parallel whenever possible
		return makePromise(() => {
			const lookaheadDepth = mapping.lookahead === LookaheadMode.PRELOAD ? mapping.lookaheadDepth || 1 : 1 // TODO - test other modes
			const lookaheadObjs = findLookaheadForlayer(playoutData, layerId, mapping.lookahead, lookaheadDepth, futureParts)
	
			// Add the objects that have some timing info
			_.each(lookaheadObjs.timed, (entry, i) => {
				let enable: TimelineTypes.TimelineEnable = {
					start: 1 // Absolute 0 without a group doesnt work
				}
				if (i !== 0) {
					const prevObj = lookaheadObjs.timed[i - 1].obj
					enable = calculateStartAfterPreviousObj(prevObj)
				}
				if (!entry.obj.id) throw new Meteor.Error(500, 'lookahead: timeline obj id not set')
	
				enable.end = `#${entry.obj.id}.start`
	
				mutateAndPushObject(entry.obj, `timed${i}`, enable, mapping, LOOKAHEAD_OBJ_PRIORITY)
			})
	
			// Add each of the future objects, that have no end point
			const futureObjCount = lookaheadObjs.future.length
			const futurePriorityScale = LOOKAHEAD_OBJ_PRIORITY / (futureObjCount + 1)
			_.each(lookaheadObjs.future, (entry, i) => {
				if (!entry.obj.id) throw new Meteor.Error(500, 'lookahead: timeline obj id not set')
	
				// WHEN_CLEAR mode can't take multiple futures, as they are always flattened into the single layer. so give it some real timings, and only output one
				const singleFutureObj = mapping.lookahead !== LookaheadMode.WHEN_CLEAR
				if (singleFutureObj && i !== 0) {
					return
				}
	
				const lastTimedObj = _.last(lookaheadObjs.timed)
				const enable = singleFutureObj && lastTimedObj ? calculateStartAfterPreviousObj(lastTimedObj.obj) : { while: '1' }
				// We use while: 1 for the enabler, as any time before it should be active will be filled by either a playing object, or a timed lookahead.
				// And this allows multiple futures to be timed in a way that allows them to co-exist
	
				// Prioritise so that the earlier ones are higher, decreasing within the range 'reserved' for lookahead
				const priority = singleFutureObj ? LOOKAHEAD_OBJ_PRIORITY : futurePriorityScale * (futureObjCount - i)
				mutateAndPushObject(entry.obj, `future${i}`, enable, mapping, priority)
			})
		})
	})
	waitForPromiseAll(ps)
	
	return timelineObjs
}

export interface LookaheadObjectEntry {
	obj: TimelineObjRundown
	partId: PartId
}

export interface LookaheadResult {
	timed: Array<LookaheadObjectEntry>
	future: Array<LookaheadObjectEntry>
}

export function findLookaheadForlayer (
	playoutData: RundownPlaylistPlayoutData,
	layer: string,
	mode: LookaheadMode,
	lookaheadDepth: number,
	futureParts: Part[],
	// followingPiecesQueryFragment: Mongo.Query<Piece>[]
	): LookaheadResult {

	const res: LookaheadResult = {
		timed: [],
		future: []
	}

	if (mode === undefined || mode === LookaheadMode.NONE) {
		return res
	}

	// Track the previous info for checking how the timeline will be built
	let previousPart: Part | undefined
	if (playoutData.previousPartInstance) {
		previousPart = playoutData.previousPartInstance.part
	}

	// Get the PieceInstances which are on the timeline
	const partInstancesOnTimeline = _.compact([
		playoutData.currentPartInstance,
		playoutData.nextPartInstance
	])
	// Generate timed objects for parts on the timeline
	_.each(partInstancesOnTimeline, partInstance => {
		const pieces = _.filter(playoutData.selectedInstancePieces, (pieceInstance: PieceInstance) => {
			return !!(
				pieceInstance.partInstanceId === partInstance._id &&
				pieceInstance.piece.content &&
				pieceInstance.piece.content.timelineObjects &&
				_.find(pieceInstance.piece.content.timelineObjects, (o) => (o && o.layer === layer))
			)
		})

		const newObjs = findObjectsForPart(playoutData.rundownPlaylist.currentPartInstanceId, layer, previousPart, partInstance.part, pieces.map(p => p.piece), partInstance)
		const isAutonext = playoutData.currentPartInstance && playoutData.currentPartInstance.part.autoNext
		if (playoutData.rundownPlaylist.nextPartInstanceId === partInstance._id && !isAutonext) {
			// The next instance should be future objects if not in an autonext
			newObjs.forEach(o => res.future.push({ obj: o, partId: partInstance.part._id }))
		} else {
			newObjs.forEach(o => res.timed.push({ obj: o, partId: partInstance.part._id }))
		}
		previousPart = partInstance.part
	})

	// Ensure we havent reached enough already
	if (res.future.length >= lookaheadDepth) {
		return res
	}
	
	// There are no future parts, so this will not find anything more
	if (futureParts.length === 0) {
		return res
	}
	const futurePartIds = _.map(futureParts, part => part._id)

	// find enough pieces that touch the layer
	const rundownIds = Object.keys(playoutData.rundownsMap).map(id => protectString<RundownId>(id))
	const piecesUsingLayer: Array<Pick<Piece, 'startPartId'>> = Pieces.find({
		invalid: { $ne: true },
		'content.timelineObjects.layer': layer,
		startRundownId: { $in: rundownIds },
		startPartId: { $in: futurePartIds }
	}, {
		limit: lookaheadDepth * 2, // TODO - is this enough because some could be dropped due to being transitions?
		fields: {
			startPartId: 1
		}
	}).fetch()

	if (piecesUsingLayer.length === 0) {
		// No instances of layer
		return res
	}

	// We need all the pieces from those parts..
	const possiblePartIds = _.uniq(_.map(piecesUsingLayer, piece => piece.startPartId))
	const allPiecesFromParts = Pieces.find({
		invalid: { $ne: true },
		startRundownId: { $in: rundownIds },
		startPartId: { $in: possiblePartIds }
	}).fetch()
	const piecesUsingLayerByPart = _.groupBy(allPiecesFromParts, piece => piece.startPartId)

	for (const part of futureParts) {
		// Stop if we have enough objects already
		if (res.future.length >= lookaheadDepth) {
			break
		}

		const pieces = piecesUsingLayerByPart[unprotectString(part._id)] || []
		if (pieces.length > 0 && part.isPlayable()) {
			findObjectsForPart(playoutData.rundownPlaylist.currentPartInstanceId, layer, previousPart, part, pieces, null)
				.forEach(o => res.future.push({ obj: o, partId: part._id }))
		}
		previousPart = part
	}

	return res
}

export function findObjectsForPart (
	currentPartInstanceId: PartInstanceId | null,
	layer: string,
	previousPart: Part | undefined,
	part: Part,
	pieces: PieceInstancePiece[],
	partInstance: PartInstance | null,
): (TimelineObjRundown & OnGenerateTimelineObj)[] {

	// Sanity check, if no part to search, then abort
	if (!part || !pieces || pieces.length === 0) {
		return []
	}

	let allObjs: TimelineObjRundown[] = []
	pieces.forEach(piece => {
		if (piece.content && piece.content.timelineObjects) {
			// Calculate the pieceInstanceId or fallback to the pieceId. This is ok, as its only for lookahead
			const pieceInstanceId = partInstance ? rewrapPieceToInstance(piece, partInstance.rundownId, partInstance._id)._id : piece._id

			_.each(piece.content.timelineObjects, (obj) => {
				if (obj) {
					fixTimelineId(obj)
					allObjs.push(literal<TimelineObjRundown & OnGenerateTimelineObj>({
						...obj,
						_id: protectString(''), // set later
						studioId: protectString(''), // set later
						objectType: TimelineObjType.RUNDOWN,
						pieceInstanceId: unprotectString(pieceInstanceId),
						infinitePieceId: unprotectString(piece.infiniteId)
					}))
				}
			})
		}
	})
	// let allObjs: TimelineObjRundown[] = _.compact(rawObjs)

	if (allObjs.length === 0) {
		// Should never happen. suggests something got 'corrupt' during this process
		return []
	} else if (allObjs.length === 1) {
		// Only one, just return it
		return allObjs
	} else { // They need to be ordered
		// TODO - this needs to consider infinites properly... In what way?
		const orderedPieces = sortPiecesByStart(pieces)

		let allowTransition = false
		let classesFromPreviousPart: string[] = []
		if (previousPart && currentPartInstanceId) { // If we have a previous and not at the start of the rundown
			allowTransition = !previousPart.disableOutTransition
			classesFromPreviousPart = previousPart.classesForNext || []
		}

		const transObj = orderedPieces.find(i => !!i.isTransition)
		const transObj2 = transObj ? pieces.find(l => l._id === transObj._id) : undefined
		const hasTransition = (
			allowTransition &&
			transObj2 &&
			transObj2.content &&
			transObj2.content.timelineObjects &&
			transObj2.content.timelineObjects.find(o => o != null && o.layer === layer)
		)

		const res: TimelineObjRundown[] = []
		orderedPieces.forEach(piece => {
			if (!part || (!allowTransition && piece.isTransition)) {
				return
			}
			if (!piece.content || !piece.content.timelineObjects) {
				return
			}

			// If there is a transition and this piece is abs0, it is assumed to be the primary piece and so does not need lookahead
			if (
				hasTransition &&
				!piece.isTransition &&
				piece.enable.start === 0 // <-- need to discuss this!
			) {
				return
			}

			// Note: This is assuming that there is only one use of a layer in each piece.
			const obj = piece.content.timelineObjects.find(o => o !== null && o.layer === layer)
			if (obj) {
				// Try and find a keyframe that is used when in a transition
				let transitionKF: TimelineTypes.TimelineKeyframe | undefined = undefined
				if (allowTransition) {
					transitionKF = _.find(obj.keyframes || [], kf => kf.enable.while === '.is_transition')

					// TODO - this keyframe matching is a hack, and is very fragile

					if (!transitionKF && classesFromPreviousPart && classesFromPreviousPart.length > 0) {
						// Check if the keyframe also uses a class to match. This handles a specific edge case
						transitionKF = _.find(obj.keyframes || [], kf => _.any(classesFromPreviousPart, cl => kf.enable.while === `.is_transition & .${cl}`))
					}
				}
				const newContent = Object.assign({}, obj.content, transitionKF ? transitionKF.content : {})

				// Calculate the pieceInstanceId or fallback to the pieceId. This is ok, as its only for lookahead
				const pieceInstanceId = partInstance ? rewrapPieceToInstance(piece, partInstance.rundownId, partInstance._id)._id : piece._id

				res.push(literal<TimelineObjRundown & OnGenerateTimelineObj>({
					...obj,
					_id: protectString(''), // set later
					studioId: protectString(''), // set later
					objectType: TimelineObjType.RUNDOWN,
					pieceInstanceId: unprotectString(pieceInstanceId),
					infinitePieceId: unprotectString(piece.infiniteId),
					content: newContent
				}))
			}
		})
		return res
	}
}
