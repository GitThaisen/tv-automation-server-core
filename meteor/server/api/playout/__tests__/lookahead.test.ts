import { Meteor } from 'meteor/meteor'
import '../../../../__mocks__/_extendJest'
import { testInFiber } from '../../../../__mocks__/helpers/jest'
import { fixSnapshot } from '../../../../__mocks__/helpers/snapshot'
import { mockupCollection } from '../../../../__mocks__/helpers/lib'
import { setupDefaultStudioEnvironment, DefaultEnvironment, setupDefaultRundownPlaylist, setupMockPeripheralDevice } from '../../../../__mocks__/helpers/database'
import { ShowStyleBase, ShowStyleBases, DBShowStyleBase } from '../../../../lib/collections/ShowStyleBases';
import { Rundowns, RundownId, DBRundown, Rundown } from '../../../../lib/collections/Rundowns';
import { Segments, SegmentId, DBSegment } from '../../../../lib/collections/Segments';
import { Parts, PartId, DBPart, Part } from '../../../../lib/collections/Parts';
import { protectString, literal, normalizeArray } from '../../../../lib/lib';
import { RundownPlaylistId, RundownPlaylistPlayoutData, DBRundownPlaylist, RundownPlaylist } from '../../../../lib/collections/RundownPlaylists';
import { Piece, PieceId, Pieces } from '../../../../lib/collections/Pieces';
import { PieceLifespan, SourceLayerType, TSR, TimelineObjectCoreExt, LookaheadMode } from 'tv-automation-sofie-blueprints-integration';
import { PartInstance, DBPartInstance, PartInstanceId } from '../../../../lib/collections/PartInstances';
import { PieceInstances, PieceInstance, PieceInstancePiece } from '../../../../lib/collections/PieceInstances';
import { findLookaheadForlayer } from '../lookahead'

const Ids = {
	playlist: protectString<RundownPlaylistId>('playlist0'),
	rundown0: protectString<RundownId>('rundown0'),
	rundown0Rank: 0,
	rundown1: protectString<RundownId>('rundown1'),
	rundown1Rank: 1,

	part0: protectString<PartId>('part0'),
	part1: protectString<PartId>('part1'),
	part2: protectString<PartId>('part2'),
	part3: protectString<PartId>('part3'),
	part4: protectString<PartId>('part4'),

	partInstance0: protectString<PartInstanceId>('partInstance0'),
	partInstance1: protectString<PartInstanceId>('partInstance1'),
	partInstance2: protectString<PartInstanceId>('partInstance2'),
	partInstance3: protectString<PartInstanceId>('partInstance3'),
}
const rundownIds = [Ids.rundown0, Ids.rundown1]
const Fakes = {
	playlist: new RundownPlaylist(literal<Partial<DBRundownPlaylist>>({
		_id: Ids.playlist,
	}) as DBRundownPlaylist),
	rundown0: new Rundown(literal<Partial<DBRundown>>({
		_id: Ids.rundown0,
		_rank: Ids.rundown0Rank
	}) as DBRundown),
	rundown1: new Rundown(literal<Partial<DBRundown>>({
		_id: Ids.rundown1,
		_rank: Ids.rundown1Rank
	}) as DBRundown),

	part0: new Part(literal<Partial<DBPart>>({
		_id: Ids.part0
	}) as DBPart),
	part1: new Part(literal<Partial<DBPart>>({
		_id: Ids.part1
	}) as DBPart),
	part2: new Part(literal<Partial<DBPart>>({
		_id: Ids.part2
	}) as DBPart),
	part3: new Part(literal<Partial<DBPart>>({
		_id: Ids.part3
	}) as DBPart),
	part4: new Part(literal<Partial<DBPart>>({
		_id: Ids.part4
	}) as DBPart),
}
const Fakes2 = {
	partInstance0: new PartInstance(literal<Partial<DBPartInstance>>({
		_id: Ids.partInstance0,
		part: Fakes.part0
	}) as DBPartInstance),
	partInstance1: new PartInstance(literal<Partial<DBPartInstance>>({
		_id: Ids.partInstance1,
		part: Fakes.part1
	}) as DBPartInstance),
	partInstance2: new PartInstance(literal<Partial<DBPartInstance>>({
		_id: Ids.partInstance2,
		part: Fakes.part2
	}) as DBPartInstance),
	partInstance3: new PartInstance(literal<Partial<DBPartInstance>>({
		_id: Ids.partInstance3,
		part: Fakes.part3
	}) as DBPartInstance),

	pieceInstances1: literal<Array<Partial<PieceInstance>>>([
		{
			_id: protectString('pieceInstance1_0'),
			partInstanceId: Ids.partInstance1,
			piece: {
				enable: { start: 0 },
				content: {
					timelineObjects: [
						{
							id: 'some-obj',
							enable: { start: 0 },
							layer: 'layer1'
						} as TimelineObjectCoreExt
					]
				}
			} as PieceInstancePiece
		},
		{
			_id: protectString('pieceInstance1_1'),
			partInstanceId: Ids.partInstance1,
			piece: {
				enable: { start: 0 },
				content: {
					timelineObjects: [
						{
							id: 'some-obj2',
							enable: { start: 0 },
							layer: 'layer2'
						} as TimelineObjectCoreExt
					]
				}
			} as PieceInstancePiece
		}
	]) as PieceInstance[],
	pieceInstances2: literal<Array<Partial<PieceInstance>>>([
		{
			_id: protectString('pieceInstance2_0'),
			partInstanceId: Ids.partInstance2,
			piece: {
				enable: { start: 0 },
				content: {
					timelineObjects: [
						{
							id: 'some-obj3',
							enable: { start: 0 },
							layer: 'layer1'
						} as TimelineObjectCoreExt
					]
				}
			} as PieceInstancePiece
		},
		{
			_id: protectString('pieceInstance2_1'),
			partInstanceId: Ids.partInstance2,
			piece: {
				enable: { start: 1000 },
				content: {
					timelineObjects: [
						{
							id: 'some-obj4',
							enable: { start: 0 },
							layer: 'layer1'
						} as TimelineObjectCoreExt
					]
				}
			} as PieceInstancePiece
		}
	]) as PieceInstance[],
}

const playoutData: RundownPlaylistPlayoutData = {
	rundownPlaylist: Fakes.playlist,
	rundowns: [Fakes.rundown0, Fakes.rundown1],
	rundownsMap: normalizeArray([Fakes.rundown0, Fakes.rundown1], '_id'),

	// TODO
	currentPartInstance: undefined,
	nextPartInstance: undefined,
	previousPartInstance: undefined,
	selectedInstancePieces: []
}

function setupBaseMockInfinitesRundown() {
	// Clean out old stuff
	ShowStyleBases.remove({ _id: { $exists: true } })
	Pieces.remove({ _id: { $exists: true } })

	playoutData.currentPartInstance = undefined
	playoutData.nextPartInstance = undefined
	playoutData.previousPartInstance = undefined
	playoutData.selectedInstancePieces = []
	playoutData.rundownPlaylist.currentPartInstanceId = null
	playoutData.rundownPlaylist.nextPartInstanceId = null
	playoutData.rundownPlaylist.previousPartInstanceId = null

	Pieces.insert({
		_id: protectString('random piece'),
		startRundownId: Ids.rundown0,
		startRundownRank: Ids.rundown0Rank,
		startSegmentId: protectString('segment0'),
		startSegmentRank: 0,
		startPartId: Ids.part3,
		startPartRank: 3,

		externalId: '',
		status: 0,
		name: '',
		lifespan: PieceLifespan.OutOnRundownEnd,
		sourceLayerId: 'layer0',
		outputLayerId: 'pgm',
		enable: { start: 0 },
		invalid: false,
		content: {
			timelineObjects: [
				{
					id: 'some-obj',
					enable: { start: 0 },
					layer: 'some-random-layer'
				} as TimelineObjectCoreExt
			]
		}
	})
	
}

describe('Lookahead Calculations', () => {
	beforeEach(() => {
		setupBaseMockInfinitesRundown()
	})
	describe('findLookaheadForlayer', () => {
		testInFiber('empty layer should produce nothing', () => {
			{
				// From no active parts
				const parts = [Fakes.part0, Fakes.part1, Fakes.part2]
				const result = findLookaheadForlayer(playoutData, 'layer99', LookaheadMode.PRELOAD, 1, parts)
				expect(result.future).toHaveLength(0)
				expect(result.timed).toHaveLength(0)
			}

			{
				// With a current part
				const playoutData2: RundownPlaylistPlayoutData = {
					...playoutData,
					currentPartInstance: Fakes2.partInstance1,
					nextPartInstance: Fakes2.partInstance2,
					selectedInstancePieces: [...Fakes2.pieceInstances1, ...Fakes2.pieceInstances2]
				}
				const parts = [Fakes.part3, Fakes.part4]
				const result = findLookaheadForlayer(playoutData2, 'layer99', LookaheadMode.PRELOAD, 1, parts)
				expect(result.future).toHaveLength(0)
				expect(result.timed).toHaveLength(0)
			}
		})
		testInFiber('layer should produce something', () => {
			Pieces.insert({
				_id: protectString('piece0'),
				startRundownId: Ids.rundown0,
				startRundownRank: Ids.rundown0Rank,
				startSegmentId: protectString('segment0'),
				startSegmentRank: 0,
				startPartId: Ids.part3,
				startPartRank: 3,

				externalId: '',
				status: 0,
				name: '',
				lifespan: PieceLifespan.OutOnRundownEnd,
				sourceLayerId: 'layer0',
				outputLayerId: 'pgm',
				enable: { start: 0 },
				invalid: false,
				content: {
					timelineObjects: [
						{
							id: 'find-this-obj',
							enable: { start: 0 },
							layer: 'layer1'
						} as TimelineObjectCoreExt
					]
				}
			})
			{
				// From no active parts
				const parts = [Fakes.part0, Fakes.part1, Fakes.part2]
				const result = findLookaheadForlayer(playoutData, 'layer1', LookaheadMode.PRELOAD, 1, parts)
				expect(result.future).toHaveLength(0)
				expect(result.timed).toHaveLength(0)
			}

			{
				// From no active parts 2
				const parts = [Fakes.part0, Fakes.part1, Fakes.part2, Fakes.part3, Fakes.part4]
				const result = findLookaheadForlayer(playoutData, 'layer1', LookaheadMode.PRELOAD, 1, parts)
				expect(result.future).toHaveLength(1)
				expect(result.timed).toHaveLength(0)
				expect(result).toMatchSnapshot()
			}

			{
				// With a current part
				const playoutData2: RundownPlaylistPlayoutData = {
					...playoutData,
					currentPartInstance: Fakes2.partInstance1,
					nextPartInstance: Fakes2.partInstance2,
					selectedInstancePieces: [...Fakes2.pieceInstances1, ...Fakes2.pieceInstances2]
				}
				const parts = [Fakes.part3, Fakes.part4]
				const result = findLookaheadForlayer(playoutData2, 'layer1', LookaheadMode.PRELOAD, 1, parts)
				expect(result.future).toHaveLength(1)
				expect(result.timed).toHaveLength(3)
				expect(result).toMatchSnapshot()
			}
		})
		
		// TODO - more tests
	})
})