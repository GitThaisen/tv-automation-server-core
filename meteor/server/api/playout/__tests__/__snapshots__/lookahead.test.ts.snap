// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lookahead Calculations findLookaheadForlayer layer should produce something 1`] = `
Object {
  "future": Array [
    Object {
      "obj": Object {
        "_id": "",
        "content": Object {},
        "enable": Object {
          "start": 0,
        },
        "id": "find-this-obj",
        "infinitePieceId": undefined,
        "layer": "layer1",
        "objectType": "rundown",
        "pieceInstanceId": "piece0",
        "studioId": "",
      },
      "partId": "part3",
    },
  ],
  "timed": Array [],
}
`;

exports[`Lookahead Calculations findLookaheadForlayer layer should produce something 2`] = `
Object {
  "future": Array [
    Object {
      "obj": Object {
        "_id": "",
        "content": Object {},
        "enable": Object {
          "start": 0,
        },
        "id": "find-this-obj",
        "infinitePieceId": undefined,
        "layer": "layer1",
        "objectType": "rundown",
        "pieceInstanceId": "piece0",
        "studioId": "",
      },
      "partId": "part3",
    },
  ],
  "timed": Array [
    Object {
      "obj": Object {
        "_id": "",
        "enable": Object {
          "start": 0,
        },
        "id": "some-obj",
        "infinitePieceId": undefined,
        "layer": "layer1",
        "objectType": "rundown",
        "pieceInstanceId": "partInstance1_undefined",
        "studioId": "",
      },
      "partId": "part1",
    },
    Object {
      "obj": Object {
        "_id": "",
        "content": Object {},
        "enable": Object {
          "start": 0,
        },
        "id": "some-obj3",
        "infinitePieceId": undefined,
        "layer": "layer1",
        "objectType": "rundown",
        "pieceInstanceId": "partInstance2_undefined",
        "studioId": "",
      },
      "partId": "part2",
    },
    Object {
      "obj": Object {
        "_id": "",
        "content": Object {},
        "enable": Object {
          "start": 0,
        },
        "id": "some-obj4",
        "infinitePieceId": undefined,
        "layer": "layer1",
        "objectType": "rundown",
        "pieceInstanceId": "partInstance2_undefined",
        "studioId": "",
      },
      "partId": "part2",
    },
  ],
}
`;
